src/command/backup/protocol.c

```diff
@@ -9,41 +9,56 @@ Backup Protocol Handler
     MEM_CONTEXT_TEMP_BEGIN()
     {
-        // Backup the file
-        BackupFileResult result = backupFile(
-            varStr(varLstGet(paramList, 0)), varBool(varLstGet(paramList, 1)), varUInt64(varLstGet(paramList, 2)),
-            varBool(varLstGet(paramList, 3)), varStr(varLstGet(paramList, 4)), varBool(varLstGet(paramList, 5)),
-            varUInt64(varLstGet(paramList, 6)), varStr(varLstGet(paramList, 7)), varBool(varLstGet(paramList, 8)),
-            (CompressType)varUIntForce(varLstGet(paramList, 9)), varIntForce(varLstGet(paramList, 10)),
-            varStr(varLstGet(paramList, 11)), varBool(varLstGet(paramList, 12)), varUInt64(varLstGet(paramList, 13)),
-            varStr(varLstGet(paramList, 14)));
-
-        // Return backup result
-        VariantList *resultList = varLstNew();
-        varLstAdd(resultList, varNewUInt(result.backupCopyResult));
-        varLstAdd(resultList, varNewUInt64(result.copySize));
-        varLstAdd(resultList, varNewUInt64(result.repoSize));
-        varLstAdd(resultList, varNewStr(result.copyChecksum));
-        varLstAdd(resultList, result.pageChecksumResult != NULL ? varNewKv(result.pageChecksumResult) : NULL);
-
-        protocolServerResponse(server, varNewVarLst(resultList));
+        // Backup file
+        const String *const pgFile = pckReadStrP(param);
+        const bool pgFileIgnoreMissing = pckReadBoolP(param);
+        const uint64_t pgFileSize = pckReadU64P(param);
+        const bool pgFileCopyExactSize = pckReadBoolP(param);
+        const String *const pgFileChecksum = pckReadStrP(param);
+        const bool pgFileChecksumPage = pckReadBoolP(param);
+        const uint64_t pgFileChecksumPageLsnLimit = pckReadU64P(param);
+        const String *const repoFile = pckReadStrP(param);
+        const bool repoFileHasReference = pckReadBoolP(param);
+        const CompressType repoFileCompressType = (CompressType)pckReadU32P(param);
+        const int repoFileCompressLevel = pckReadI32P(param);
+        const String *const backupLabel = pckReadStrP(param);
+        const bool delta = pckReadBoolP(param);
+        const CipherType cipherType = (CipherType)pckReadU64P(param);
+        const String *const cipherPass = pckReadStrP(param);
+
+        const BackupFileResult result = backupFile(
+            pgFile, pgFileIgnoreMissing, pgFileSize, pgFileCopyExactSize, pgFileChecksum, pgFileChecksumPage,
+            pgFileChecksumPageLsnLimit, repoFile, repoFileHasReference, repoFileCompressType, repoFileCompressLevel,
+            backupLabel, delta, cipherType, cipherPass);
+
+        // Return result
+        PackWrite *const resultPack = protocolPackNew();
+        pckWriteU32P(resultPack, result.backupCopyResult);
+        pckWriteU64P(resultPack, result.copySize);
+        pckWriteU64P(resultPack, result.repoSize);
+        pckWriteStrP(resultPack, result.copyChecksum);
+        pckWriteStrP(resultPack, result.pageChecksumResult != NULL ? jsonFromKv(result.pageChecksumResult) : NULL);
```

Comment:
When explicitly declaring NULL here, can I assume the NULL will not be written and will result in a skipped field ID (i.e. when result.pageChecksumResult is NULL then pckWriteStrP(resultPack, NULL) is "written"....but not really... it will just skipped because (from pack.h) `NULLs are not stored in a pack and are therefore not typed. A NULL is essentially just a gap in the field IDs. `

So the pckReadStr in backup.c we explicitly set .defaultValue = NULL_STR (which is an actual character representation "null") when we encounter the skipping of the field ID because `When using read functions the default will always be returned when the field is NULL (i.e. missing). The standard default is the C default for that type (e.g. bool = false, int = 0) but can be changed with the .defaultValue parameter.`

There are a couple typos in the comments of pack.h so I wanted to get clarification on this example before I made any changes...

Reply:
> When explicitly declaring NULL here, can I assume the NULL will not be written and will result in a skipped field ID

Yes.

> So the pckReadStr in backup.c we explicitly set .defaultValue = NULL_STR (which is an actual character representation "null") when we encounter the skipping of the field ID

Yes.

This would be a cleaner if jsonFromKv() accepted NULL. We can make that change if you like.

Reply:
Updated the comments in pack.h instead in commit 0f1c2eb4ee4674d7c22716ae3f5f0420cdc6592f